caddyhttp/staticfiles/fileserver.go

```diff
@@ -80,17 +80,37 @@ func (fs FileServer) serveFile(w http.ResponseWriter, r *http.Request, name stri
 	}
 
 	// redirect to canonical path
-	url := r.URL.Path
+	u := r.URL.Path
 	if d.IsDir() {
 		// Ensure / at end of directory url
-		if !strings.HasSuffix(url, "/") {
-			Redirect(w, r, path.Base(url)+"/", http.StatusMovedPermanently)
+		if !strings.HasSuffix(u, "/") {
+			// Use the request's RequestURI since the path in u may have been
+			// trimmed by Caddy.
+			redirURL, err := url.ParseRequestURI(r.RequestURI)
+			if err != nil {
+				return http.StatusInternalServerError, err
+			}
+			if !strings.HasSuffix(redirURL.Path, "/") {
+				redirURL.Path = redirURL.Path + "/"
+			}
+			toURL := r.URL.ResolveReference(redirURL)
+
+			http.Redirect(w, r, toURL.String(), http.StatusMovedPermanently)
```

Comment:
My concern with changing fields in r.URL is that it might cause issues for other middleware Handlers that rely on them. Otherwise, only the RequestURI seems to have the original path. Let me know if that's not the case.

The check for "/" on redirURL.Path is probably redundant and can be removed.

I'll make that change and add the requested comment. Thanks for taking a look!